flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/queue/UnorderedStreamElementQueue.java

```diff
@@ -218,88 +172,99 @@ public int size() {
 	}
 
 	/**
-	 * Callback for onComplete events for the given stream element queue entry. Whenever a queue
-	 * entry is completed, it is checked whether this entry belongs to the first set. If this is the
-	 * case, then the element is added to the completed entries queue from where it can be consumed.
-	 * If the first set becomes empty, then the next set is polled from the uncompleted entries
-	 * queue. Completed entries from this new set are then added to the completed entries queue.
-	 *
-	 * @param streamElementQueueEntry which has been completed
-	 * @throws InterruptedException if the current thread has been interrupted while performing the
-	 * 	on complete callback.
+	 * An entry that notifies the respective stage upon completion.
 	 */
-	public void onCompleteHandler(StreamElementQueueEntry<?> streamElementQueueEntry) throws InterruptedException {
-		lock.lockInterruptibly();
+	static class StagedStreamRecordQueueEntry<OUT> extends StreamRecordQueueEntry<OUT> {
+		private final Stage stage;
 
-		try {
-			if (firstSet.remove(streamElementQueueEntry)) {
-				completedQueue.offer(streamElementQueueEntry);
+		public StagedStreamRecordQueueEntry(StreamRecord<?> inputRecord, Stage stage) {
+			super(inputRecord);
+			this.stage = stage;
+		}
 
-				while (firstSet.isEmpty() && firstSet != lastSet) {
-					firstSet = uncompletedQueue.poll();
+		@Override
+		public void complete(Collection<OUT> result) {
+			super.complete(result);
+			this.stage.completed(this);
+		}
+	}
 
-					Iterator<StreamElementQueueEntry<?>> it = firstSet.iterator();
+	/**
+	 * A stage is a collection of queue entries that can be completed in arbitrary order.
+	 */
+	static class Stage<OUT> {
+		/** Undrained finished elements. */
+		private final Queue<StreamElementQueueEntry<OUT>> completedQueue;
 
-					while (it.hasNext()) {
-						StreamElementQueueEntry<?> bufferEntry = it.next();
+		/** Finished and unfinished input elements. Used solely for checkpointing. */
+		private final Set<StreamElementQueueEntry<OUT>> pendingElements;
```

Comment:
I think it is better to rename `uncompletedElements` instead, because the semantic `pending` in below `addPendingElements` covers both completed/uncompleted elements.